File: ui/components/status_indicator.py
# -*- coding: utf-8 -*-
"""
Status effect visual indicators for combat
"""

from PySide6.QtWidgets import QLabel, QFrame, QHBoxLayout, QVBoxLayout
from PySide6.QtCore import Qt, QTimer, QEasingCurve, Signal
from PySide6.QtGui import QFont


class StatusIndicator(QFrame):
    """Visual indicator for status effects like buffs/debuffs"""

    def __init__(self, effect_type, effect_name, duration=None, parent=None):
        super().__init__(parent)
        self.effect_type = effect_type
        self.effect_name = effect_name
        self.duration = duration
        self.remaining_time = duration

        self.init_ui()

        # Set up timer for duration countdown if applicable
        if duration and duration > 0:
            self.timer = QTimer(self)
            self.timer.timeout.connect(self.update_duration)
            self.timer.start(1000)  # Update every second

    def update_duration(self):
        """Update the remaining duration and handle expiration"""
        if self.remaining_time is not None:
            self.remaining_time -= 1
            if self.remaining_time <= 0:
                self.expire_effect()
            else:
                self.update_display()

    def update_display(self):
        """Update the visual display of the status indicator"""
        # This can be overridden by subclasses for custom display updates
        pass

    def expire_effect(self):
        """Handle effect expiration"""
        if hasattr(self, 'timer'):
            self.timer.stop()
        self.hide()
        # Emit signal or handle cleanup as needed

    def get_effect_colors(self):
        """Get colors based on effect type"""
        color_map = {
            'buff': {'bg': '#d4edda', 'border': '#27ae60'},
            'debuff': {'bg': '#f8d7da', 'border': '#e74c3c'},
            'heal': {'bg': '#d1ecf1', 'border': '#3498db'},
            'damage': {'bg': '#fff3cd', 'border': '#f39c12'},
            'default': {'bg': '#f8f9fa', 'border': '#6c757d'}
        }
        return color_map.get(self.effect_type, color_map['default'])

    def get_effect_icon(self):
        """Get icon/emoji based on effect type"""
        icon_map = {
            'buff': '💪',
            'debuff': '☠️',
            'heal': '❤️',
            'damage': '💥',
            'poison': '☠️',
            'strength': '💪',
            'defense': '🛡️',
            'speed': '⚡',
            'regeneration': '💚',
            'burn': '🔥',
            'freeze': '🧊',
            'stun': '💫',
            'default': '✨'
        }
        return icon_map.get(self.effect_type, icon_map['default'])

    def init_ui(self):
        """Initialize the status indicator UI"""
        self.setFixedSize(60, 60)
        self.setFrameShape(QFrame.Shape.Box)

        # Get colors based on effect type
        colors = self.get_effect_colors()

        self.setStyleSheet(f"""
            StatusIndicator {{
                background-color: {colors['bg']};
                border: 2px solid {colors['border']};
                border-radius: 8px;
            }}
        """)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(4, 4, 4, 4)
        layout.setSpacing(2)

        # Icon/emoji
        icon_label = QLabel(self.get_effect_icon())
        icon_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        icon_label.setStyleSheet("font-size: 16px;")
        layout.addWidget(icon_label)

        # Name (abbreviated if needed)

File: cpp-qt-rpg/src/components/StatusIndicator.cpp
#include "StatusIndicator.h"
#include <QLabel>
#include <QVBoxLayout>
#include <QDebug>

StatusIndicator::StatusIndicator(const QString &effectType, const QString &effectName, int duration, QWidget *parent)
    : QFrame(parent),
      m_effectType(effectType),
      m_effectName(effectName),
      m_remainingTime(duration),
      m_timer(nullptr)
{
    initUi();

    // Set up timer for duration countdown if applicable
    if (duration > 0) {
        m_timer = new QTimer(this);
        connect(m_timer, &QTimer::timeout, this, &StatusIndicator::updateDuration);
        m_timer->start(1000); // Update every second
    }
}

QMap<QString, QColor> StatusIndicator::getEffectColors() const
{
    QMap<QString, QColor> colorMap;
    colorMap["buff_bg"] = QColor("#d4edda");
    colorMap["buff_border"] = QColor("#27ae60");
    colorMap["debuff_bg"] = QColor("#f8d7da");
    colorMap["debuff_border"] = QColor("#e74c3c");
    colorMap["heal_bg"] = QColor("#d1ecf1");
    colorMap["heal_border"] = QColor("#3498db");
    colorMap["damage_bg"] = QColor("#fff3cd");
    colorMap["damage_border"] = QColor("#f39c12");
    colorMap["default_bg"] = QColor("#f8f9fa");
    colorMap["default_border"] = QColor("#6c757d");

    QString type = m_effectType;
    if (type == "strength" || type == "defense" || type == "speed" || type == "regeneration") type = "buff";
    if (type == "poison" || type == "burn" || type == "freeze" || type == "stun") type = "debuff";

    if (colorMap.contains(type + "_bg")) {
        return {{"bg", colorMap.value(type + "_bg")}, {"border", colorMap.value(type + "_border")}};
    }
    return {{"bg", colorMap.value("default_bg")}, {"border", colorMap.value("default_border")}};
}

QString StatusIndicator::getEffectIcon() const
{
    QMap<QString, QString> iconMap;
    iconMap["buff"] = "💪";
    iconMap["debuff"] = "☠️";
    iconMap["heal"] = "❤️";
    iconMap["damage"] = "💥";
    iconMap["poison"] = "☠️";
    iconMap["strength"] = "💪";
    iconMap["defense"] = "🛡️";
    iconMap["speed"] = "⚡";
    iconMap["regeneration"] = "💚";
    iconMap["burn"] = "🔥";
    iconMap["freeze"] = "🧊";
    iconMap["stun"] = "💫";
    iconMap["default"] = "✨";

    return iconMap.value(m_effectType, iconMap["default"]);
}

void StatusIndicator::initUi()
{
    setFixedSize(60, 60);
    setFrameShape(QFrame::Box);

    // Get colors based on effect type
    QMap<QString, QColor> colors = getEffectColors();

    setStyleSheet(QString(R"(
        StatusIndicator {
            background-color: %1;
            border: 2px solid %2;
            border-radius: 8px;
        }
    )").arg(colors["bg"].name())
      .arg(colors["border"].name()));

    QVBoxLayout *layout = new QVBoxLayout(this);
    layout->setContentsMargins(4, 4, 4, 4);
    layout->setSpacing(2);

    // Icon/emoji
    QLabel *iconLabel = new QLabel(getEffectIcon());
    iconLabel->setAlignment(Qt::AlignCenter);
    iconLabel->setStyleSheet("font-size: 16px;");
    layout->addWidget(iconLabel);

    // Duration/Name Label (for display updates)
    QLabel *nameLabel = new QLabel(m_effectName);
    nameLabel->setAlignment(Qt::AlignCenter);
    nameLabel->setStyleSheet("font-size: 10px; font-weight: bold;");
    layout->addWidget(nameLabel);

    updateDisplay();
}

void StatusIndicator::updateDuration()
{
    if (m_remainingTime > 0) {
        m_remainingTime -= 1;
        if (m_remainingTime <= 0) {
            expireEffect();
        } else {
            updateDisplay();
        }
    }
}

void StatusIndicator::updateDisplay()
{
    // Find the name label (second widget in the layout)
    if (layout() && layout()->count() > 1) {
        QLabel *nameLabel = qobject_cast<QLabel*>(layout()->itemAt(1)->widget());
        if (nameLabel) {
            if (m_remainingTime > 0) {
                nameLabel->setText(QString::number(m_remainingTime) + "T"); // T for Turns
            } else {
                nameLabel->setText(m_effectName);
            }
        }
    }
}

void StatusIndicator::expireEffect()
{
    if (m_timer) {
        m_timer->stop();
        delete m_timer;
        m_timer = nullptr;
    }
    hide();
    emit effectExpired(m_effectName);
    // Note: We use deleteLater() in the Python version, which is good practice for self-destructing widgets.
    deleteLater();
}
